File: api-go/internal/services/drive_zip_test.go
package services

import (
	"archive/zip"
	"bytes"
	"context"
	"io"
	"os"
	"path/filepath"
	"testing"
)

// TestZipFolder_RoundTrip ensures the streaming zip output is a valid archive
// and contains all descendants with correct relative paths and bytes.
func TestZipFolder_RoundTrip(t *testing.T) {
	_, drive, _, _ := setupDriveFullDB(t)
	ctx := context.Background()

	root, err := drive.CreateFolder(ctx, nil, "root")
	if err != nil {
		t.Fatalf("root: %v", err)
	}
	sub, err := drive.CreateFolder(ctx, &root.ID, "sub")
	if err != nil {
		t.Fatalf("sub: %v", err)
	}
	deep, err := drive.CreateFolder(ctx, &sub.ID, "deep")
	if err != nil {
		t.Fatalf("deep: %v", err)
	}

	mustFile(t, drive, &root.ID, "a.txt", []byte("hello"))
	mustFile(t, drive, &sub.ID, "b.txt", []byte("world"))
	mustFile(t, drive, &deep.ID, "c.bin", []byte{1, 2, 3})

	var buf bytes.Buffer
	if err := drive.ZipFolder(ctx, root.ID, &buf); err != nil {
		t.Fatalf("zip: %v", err)
	}

	zr, err := zip.NewReader(bytes.NewReader(buf.Bytes()), int64(buf.Len()))
	if err != nil {
		t.Fatalf("invalid zip: %v (size=%d)", err, buf.Len())
	}

	got := map[string]string{}
	for _, f := range zr.File {
		rc, err := f.Open()
		if err != nil {
			t.Fatalf("open %s: %v", f.Name, err)
		}
		b, _ := io.ReadAll(rc)
		rc.Close()
		got[f.Name] = string(b)
	}

	want := map[string]string{
		"a.txt":          "hello",
		"sub/":           "",
		"sub/b.txt":      "world",
		"sub/deep/":      "",
		"sub/deep/c.bin": string([]byte{1, 2, 3}),
	}
	for k, v := range want {
		gv, ok := got[k]
		if !ok {
			t.Errorf("missing entry %q in zip; got entries: %v", k, mapKeys(got))
			continue
		}
		if gv != v {
			t.Errorf("entry %q: got %q, want %q", k, gv, v)
		}
	}
}

func mapKeys(m map[string]string) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	return out
}

func mustFile(t *testing.T, d *DriveService, parent *int64, name string, body []byte) {
	t.Helper()
	ctx := context.Background()
	blobRel := "test_" + name
	abs := d.BlobAbsPath(blobRel)
	if err := os.MkdirAll(filepath.Dir(abs), 0o755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}
	if err := os.WriteFile(abs, body, 0o644); err != nil {
		t.Fatalf("write blob: %v", err)
	}
	if _, err := d.CreateFileNode(ctx, parent, name, blobRel, "", int64(len(body))); err != nil {
		t.Fatalf("create file %s: %v", name, err)
	}
}

// mustFileNode is mustFile with a distinct blob name + returned node, for
// tests that need unique content per node.
func mustFileNode(t *testing.T, d *DriveService, parent *int64, name, blobRel string, body []byte) int64 {
	t.Helper()
	ctx := context.Background()
	abs := d.BlobAbsPath(blobRel)
	if err := os.MkdirAll(filepath.Dir(abs), 0o755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}
	if err := os.WriteFile(abs, body, 0o644); err != nil {
		t.Fatalf("write blob: %v", err)
	}
	n, err := d.CreateFileNode(ctx, parent, name, blobRel, "", int64(len(body)))
	if err != nil {
		t.Fatalf("create file %s: %v", name, err)
	}
	return n.ID
}

func readZip(t *testing.T, buf *bytes.Buffer) map[string]string {
	t.Helper()
	zr, err := zip.NewReader(bytes.NewReader(buf.Bytes()), int64(buf.Len()))
	if err != nil {
		t.Fatalf("invalid zip: %v (size=%d)", err, buf.Len())
	}
	got := map[string]string{}
	for _, f := range zr.File {
		rc, err := f.Open()
		if err != nil {
			t.Fatalf("open %s: %v", f.Name, err)
		}
		b, _ := io.ReadAll(rc)
		rc.Close()
		got[f.Name] = string(b)
	}
	return got
}

// ZipNodes: a mixed folder + file selection lands as top-level entries, with
// the folder keeping its own name as the top-level directory.
func TestZipNodes_MixedSelection(t *testing.T) {
	_, drive, _, _ := setupDriveFullDB(t)
	ctx := context.Background()

	folder, _ := drive.CreateFolder(ctx, nil, "photos")
	mustFile(t, drive, &folder.ID, "a.jpg", []byte("aaa"))
	rootFile := mustFileNode(t, drive, nil, "notes.txt", "test_notes.bin", []byte("nnn"))

	var buf bytes.Buffer
	if err := drive.ZipNodes(ctx, []int64{folder.ID, rootFile}, &buf); err != nil {
		t.Fatalf("zip: %v", err)
	}
	got := readZip(t, &buf)

	want := map[string]string{
		"photos/":      "",
		"photos/a.jpg": "aaa",
		"notes.txt":    "nnn",
	}
	for k, v := range want {
		if gv, ok := got[k]; !ok || gv != v {
			t.Errorf("entry %q = %q/%v, want %q; all: %v", k, gv, ok, v, mapKeys(got))
		}
	}
	if len(got) != len(want) {
		t.Errorf("entries = %d, want %d: %v", len(got), len(want), mapKeys(got))
	}
}

// ZipNodes: ids nested under other selected folders are skipped, not doubled.
func TestZipNodes_SkipsNestedSelection(t *testing.T) {
	_, drive, _, _ := setupDriveFullDB(t)
	ctx := context.Background()

	outer, _ := drive.CreateFolder(ctx, nil, "outer")
	inner, _ := drive.CreateFolder(ctx, &outer.ID, "inner")
	nestedFile := mustFileNode(t, drive, &inner.ID, "deep.txt", "test_deep.bin", []byte("ddd"))

	var buf bytes.Buffer
	if err := drive.ZipNodes(ctx, []int64{outer.ID, inner.ID, nestedFile}, &buf); err != nil {
		t.Fatalf("zip: %v", err)
	}
	got := readZip(t, &buf)

	want := map[string]string{
		"outer/":               "",
		"outer/inner/":         "",
		"outer/inner/deep.txt": "ddd",
	}
	if len(got) != len(want) {
		t.Fatalf("entries = %v, want exactly %v", mapKeys(got), mapKeys(want))
	}
	for k, v := range want {
		if got[k] != v {
			t.Errorf("entry %q = %q, want %q", k, got[k], v)
		}
	}
}

// ZipNodes: same-named top-level picks (possible from search-result
// selections) get suffixed instead of silently dropped.
func TestZipNodes_DuplicateTopLevelSuffixed(t *testing.T) {
	_, drive, _, _ := setupDriveFullDB(t)
	ctx := context.Background()

	f1, _ := drive.CreateFolder(ctx, nil, "one")
	f2, _ := drive.CreateFolder(ctx, nil, "two")
	a := mustFileNode(t, drive, &f1.ID, "dup.txt", "test_dup_a.bin", []byte("first"))
	b := mustFileNode(t, drive, &f2.ID, "dup.txt", "test_dup_b.bin", []byte("second"))

	var buf bytes.Buffer
	if err := drive.ZipNodes(ctx, []int64{a, b}, &buf); err != nil {
		t.Fatalf("zip: %v", err)
	}
	got := readZip(t, &buf)

	if got["dup.txt"] != "first" || got["dup (1).txt"] != "second" {
		t.Fatalf("entries = %v", got)
	}
}

File: api-rs/src/tests/service/drive_zip_test.rs
#[cfg(test)]
mod tests {
    use mote::config::UploadConfig;
    use mote::service::drive_service::{DriveError, DriveService};
    use mote::service::drive_zip_service::DriveZipService;
    use sqlx::sqlite::SqlitePoolOptions;
    use std::collections::HashMap;

    async fn setup() -> (
        std::sync::Arc<DriveService>,
        std::sync::Arc<DriveZipService>,
        tempfile::TempDir,
    ) {
        let tmp = tempfile::tempdir().unwrap();
        let pool = SqlitePoolOptions::new()
            .max_connections(5)
            .connect("sqlite::memory:")
            .await
            .unwrap();
        sqlx::migrate!("./migrations").run(&pool).await.unwrap();
        let cfg = UploadConfig {
            base_url: "/uploads".into(),
            base_path: tmp.path().to_string_lossy().into(),
            accel_redirect_prefix: String::new(),
            thumb_width: 240,
            image_formats: vec![],
        };
        let drive = DriveService::new(pool, cfg);
        let zip = DriveZipService::new(drive.clone());
        (drive, zip, tmp)
    }

    /// Writes a blob to disk and inserts a file node referencing it; returns the id.
    async fn mk_file(
        drive: &DriveService,
        parent_id: Option<i64>,
        name: &str,
        blob_rel: &str,
        body: &[u8],
    ) -> i64 {
        let abs = drive.blob_abs_path(blob_rel);
        std::fs::create_dir_all(abs.parent().unwrap()).unwrap();
        std::fs::write(&abs, body).unwrap();
        drive
            .create_file_node(parent_id, name, blob_rel, "", body.len() as i64)
            .await
            .unwrap()
            .id
    }

    /// Runs zip_nodes over the resolved targets and returns entry name → bytes.
    async fn zip_to_entries(zip: &DriveZipService, ids: &[i64]) -> HashMap<String, Vec<u8>> {
        let targets = zip.resolve_zip_targets(ids).await.unwrap();
        let (writer, mut reader) = tokio::io::duplex(8 << 20);
        zip.zip_nodes(&targets, writer).await.unwrap();
        let mut buf = Vec::new();
        tokio::io::AsyncReadExt::read_to_end(&mut reader, &mut buf)
            .await
            .unwrap();
        read_zip(buf).await
    }

    async fn read_zip(buf: Vec<u8>) -> HashMap<String, Vec<u8>> {
        let reader = async_zip::base::read::mem::ZipFileReader::new(buf)
            .await
            .expect("invalid zip");
        let mut out = HashMap::new();
        for i in 0..reader.file().entries().len() {
            let name = reader.file().entries()[i]
                .filename()
                .as_str()
                .unwrap()
                .to_string();
            let mut er = reader.reader_with_entry(i).await.unwrap();
            let mut data = Vec::new();
            er.read_to_end_checked(&mut data).await.unwrap();
            out.insert(name, data);
        }
        out
    }

    // A mixed folder + file selection lands as top-level entries, with the
    // folder keeping its own name as the top-level directory.
    #[tokio::test]
    async fn zip_nodes_mixed_selection() {
        let (drive, zip, _tmp) = setup().await;

        let folder = drive.create_folder(None, "photos").await.unwrap();
        mk_file(&drive, Some(folder.id), "a.jpg", "drive/test_a.bin", b"aaa").await;
        let root_file = mk_file(&drive, None, "notes.txt", "drive/test_notes.bin", b"nnn").await;

        let got = zip_to_entries(&zip, &[folder.id, root_file]).await;

        let want: &[(&str, &[u8])] = &[
            ("photos/", b""),
            ("photos/a.jpg", b"aaa"),
            ("notes.txt", b"nnn"),
        ];
        for (k, v) in want {
            assert_eq!(
                got.get(*k).map(|b| b.as_slice()),
                Some(*v),
                "entry {:?}; all: {:?}",
                k,
                got.keys().collect::<Vec<_>>()
            );
        }
        assert_eq!(
            got.len(),
            want.len(),
            "all: {:?}",
            got.keys().collect::<Vec<_>>()
        );
    }

    // Ids nested under other selected folders are skipped, not doubled.
    #[tokio::test]
    async fn zip_nodes_skips_nested_selection() {
        let (drive, zip, _tmp) = setup().await;

        let outer = drive.create_folder(None, "outer").await.unwrap();
        let inner = drive.create_folder(Some(outer.id), "inner").await.unwrap();
        let nested_file = mk_file(
            &drive,
            Some(inner.id),
            "deep.txt",
            "drive/test_deep.bin",
            b"ddd",
        )
        .await;

        let got = zip_to_entries(&zip, &[outer.id, inner.id, nested_file]).await;

        let want: &[(&str, &[u8])] = &[
            ("outer/", b""),
            ("outer/inner/", b""),
            ("outer/inner/deep.txt", b"ddd"),
        ];
        assert_eq!(
            got.len(),
            want.len(),
            "all: {:?}",
            got.keys().collect::<Vec<_>>()
        );
        for (k, v) in want {
            assert_eq!(got.get(*k).map(|b| b.as_slice()), Some(*v), "entry {:?}", k);
        }
    }

    // Same-named top-level picks (possible from search-result selections) get
    // suffixed instead of silently dropped.
    #[tokio::test]
    async fn zip_nodes_duplicate_top_level_suffixed() {
        let (drive, zip, _tmp) = setup().await;

        let f1 = drive.create_folder(None, "one").await.unwrap();
        let f2 = drive.create_folder(None, "two").await.unwrap();
        let a = mk_file(
            &drive,
            Some(f1.id),
            "dup.txt",
            "drive/test_dup_a.bin",
            b"first",
        )
        .await;
        let b = mk_file(
            &drive,
            Some(f2.id),
            "dup.txt",
            "drive/test_dup_b.bin",
            b"second",
        )
        .await;

        let got = zip_to_entries(&zip, &[a, b]).await;

        assert_eq!(
            got.get("dup.txt").map(|b| b.as_slice()),
            Some(b"first".as_slice())
        );
        assert_eq!(
            got.get("dup (1).txt").map(|b| b.as_slice()),
            Some(b"second".as_slice())
        );
    }

    // A fully deleted/missing selection resolves to NotFound before any body
    // byte would be written.
    #[tokio::test]
    async fn zip_nodes_empty_selection_not_found() {
        let (drive, zip, _tmp) = setup().await;
        let folder = drive.create_folder(None, "gone").await.unwrap();
        drive.soft_delete(&[folder.id]).await.unwrap();

        let err = zip
            .resolve_zip_targets(&[folder.id, 9999])
            .await
            .unwrap_err();
        assert!(matches!(err, DriveError::NotFound));
    }
}
